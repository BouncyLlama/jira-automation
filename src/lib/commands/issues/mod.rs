mod list_transition;
mod search_issues;
mod transition_issue;
mod update;

pub use list_transition::*;
pub use search_issues::*;
use serde::{Deserialize, Serialize};
pub use transition_issue::*;
pub use update::*;

const ISSUE_NAME_HELP: &str = "the ticket name ex. FOO-1234";
const FIX_VERSION_HELP: &str = "the version which resolves this issue";

const RELATED_VERSION_HELP: &str = "the version this issue is related to";

const TRANSITION_HELP: &str = "the name or id of transition to set";
const INCLUDE_UNAVAILABLE_HELP: &str =
    "include transitions which are not currently possible for this issue";
const BY_ID_HELP: &str = "perform the operation by specifying an ID rather than a name";

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Transition {
    id: String,
    name: String,
    is_available: bool,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IssueStatus {
    description: Option<String>,
    id: String,
    name: String,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IssueFields {
    summary: String,
    status: IssueStatus,
    description: Option<IssueDescription>,
    fix_versions: Vec<IssueRelease>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IssueDescription {
    pub(crate) content: Option<Vec<IssueDescriptionContent>>,

}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IssueDescriptionContent {
    pub(crate) content: Option<Vec<IssueDescriptionContentContent>>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IssueDescriptionContentContent {
    pub(crate) text: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct IssueRelease {
    id: String,
    description: Option<String>,
    name: String,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Issue {
    id: String,
    pub(crate) key: String,
    fields: IssueFields,
}

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PaginatedIssues {
    pub(crate) total: u64,
    pub(crate) start_at: u64,
    pub(crate) issues: Vec<Issue>,
}
