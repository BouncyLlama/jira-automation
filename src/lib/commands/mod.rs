pub mod compositions;
pub mod issues;
pub mod releases;
