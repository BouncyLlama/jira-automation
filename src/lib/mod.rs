pub mod commands;

pub mod util;

use thiserror::Error;

#[derive(Error, Debug)]
pub enum AppError {

    #[error("failed calling jira")]
    ApiCallFailed(#[from] reqwest::Error),
    #[error("failed calling jira {0}")]
    ApiCallBadStatus(String),
    #[error("IO error")]
    IOError(#[from] std::io::Error),
    #[error("failed deserializing response")]
    DeserializationError,

    #[error("query was meant to match exactly one release but found multiple")]
    MatchedMultipleReleases,
    #[error("could not create the requested release")]
    CouldNotCreateRelease,
    #[error("the specified issue transition is unknown")]
    UnknownTransition,
    #[error("no issues were found to release")]
    NoIssuesFound,
}
