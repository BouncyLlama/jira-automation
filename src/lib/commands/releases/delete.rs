use std::collections::HashMap;
use crate::Cli;
use crate::lib::util;
use clap::Parser;
use super::*;

#[derive(Parser, Clone)]
#[command()]
pub struct DeleteReleaseArgs {
    #[arg(long, short, help = RELEASE_HELP)]
    pub(crate) release: String,
    #[arg(long, short, help = PROJECT_HELP)]
    pub(crate) project: String,
    #[arg(long, short, help = BY_ID_HELP)]
    pub(crate) by_id: Option<bool>,
    #[arg(long, help = "for tickets referencing this release, replace fix version with this other release")]
    pub(crate) replace_fix_version: Option<String>,
    #[arg(long, help = "for tickets referencing this release, replace affected version with this other release")]
    pub(crate) replace_affected_version: Option<String>,

}

impl DeleteReleaseArgs {
    fn set_release(&mut self, new: String) {
        self.release = new;
    }
    fn set_fixversion(&mut self, new: String) {
        self.replace_fix_version = Option::from(new);
    }
    fn set_affectedversion(&mut self, new: String) {
        self.replace_affected_version = Option::from(new);
    }
}

pub fn execute_delete_release(ctx: &Cli, args: &DeleteReleaseArgs) -> Result<(), Box<dyn std::error::Error>> {
    let mut mutargs = args.clone();
    if args.by_id.is_some() && args.by_id.unwrap() == true {
        let req_url = format!("{}/rest/api/3/version/{}/removeAndSwap", ctx.base_jira_url, args.release);

        util::do_post::<(), HashMap<&str, String>>(&req_url, ctx, &(assemble_delete_args(args.clone())))?;
    } else {
        let id = get_id_from_name(ctx, args.project.clone(), args.release.clone())?;
        let req_url = format!("{}/rest/api/3/version/{}/removeAndSwap", ctx.base_jira_url, id);

        debug!("found release {} for name {}",id,args.release);
        mutargs.set_release(id);
        if args.replace_fix_version.is_some() {
            let fixid = get_id_from_name(ctx, args.project.clone(), args.replace_fix_version.as_ref().unwrap().clone())?;
            mutargs.set_fixversion(fixid);
        }
        if args.replace_affected_version.is_some() {
            let affectedid = get_id_from_name(ctx, args.project.clone(), args.replace_affected_version.as_ref().unwrap().clone())?;
            mutargs.set_affectedversion(affectedid);
        }
        util::do_post::<(), HashMap<&str, String>>(&req_url, ctx, &(assemble_delete_args(mutargs.clone())))?;
    }

    Ok(())
}

fn assemble_delete_args(args: DeleteReleaseArgs) -> HashMap<&'static str, String> {
    let mut params: HashMap<&str, String> = HashMap::new();
    if args.replace_affected_version.is_some() {
        params.insert("moveAffectedIssuesTo", args.replace_affected_version.unwrap());
    }
    if args.replace_fix_version.is_some() {
        params.insert("moveFixIssuesTo", args.replace_fix_version.unwrap());
    }

    return params;
}