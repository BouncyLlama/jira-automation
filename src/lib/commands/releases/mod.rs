use std::fmt::Write;
use super::super::*;
use clap::Parser;
use log::error;
use clap::Subcommand;
use clap::ArgGroup;
use log::info;
use log::warn;
use log::trace;
use log::{debug, LevelFilter};
use clap::arg;
use crate::lib::util;
use serde::Serialize;
use serde::Deserialize;
use std::{io, thread};
use std::borrow::Borrow;
use std::collections::HashMap;
use std::io::BufRead;
use std::time::Duration;
use crate::Cli;

mod test_releases;

const projectHelp: &str = "project identifier";
const nameHelp: &str = "name of the release";
const descriptionHelp: &str = "description of the release";
const startDateHelp: &str = "start date of the version in ISO 8601 format (yyyy-mm-dd)";
const releaseDateHelp: &str = "release date of the version in ISO 8601 format (yyyy-mm-dd)";
const byIdHelp: &str = "perform operation by specifying id rather than name (useful if your names are not unique)";

#[derive(Parser, Clone)]
#[command()]
pub struct ListReleasesArgs {
    #[arg(long, short, help = projectHelp)]
    pub(crate) project: String,
    #[arg(long, short, help = "optionally filter results; substring match in the name and description fields")]
    pub(crate) filter: Option<String>,
    #[arg(long, short, default_value_t = false, help = "automatically query until all pages have been obtained")]
    unpaginate: bool,
    #[arg(long, default_value_t = 50, help = "how many items to return")]
    page_size: i64,
    #[arg(long, default_value_t = 0, help = "item index to begin paging at")]
    page_start_idx: i64,
}

#[derive(Parser, Clone)]
#[command()]
pub struct CreateReleaseArgs {
    #[arg(long, short, help = projectHelp)]
    pub(crate) project: String,
    #[arg(long, short, help = nameHelp)]
    pub(crate) name: String,
    #[arg(long, short, help = descriptionHelp)]
    pub(crate) description: Option<String>,
    #[arg(long, short, help = startDateHelp)]
    pub(crate) startDate: Option<String>,
    #[arg(long, short, help = releaseDateHelp)]
    pub(crate) releaseDate: Option<String>,
}

#[derive(Parser, Clone)]
#[command()]
pub struct UpdateReleaseArgs {
    #[arg(long, short, help = nameHelp)]
    pub(crate) name: Option<String>,
    #[arg(long, short, help = descriptionHelp)]
    pub(crate) description: Option<String>,
    #[arg(long, help = startDateHelp)]
    pub(crate) startDate: Option<String>,
    #[arg(long, help = releaseDateHelp)]
    pub(crate) releaseDate: Option<String>,
    #[arg(long, short, help = "whether or not this release has been 'released'")]
    pub(crate) is_released: Option<bool>,
    #[arg(long, short, help = byIdHelp)]
    pub(crate) byId: Option<bool>,
    #[arg(long, short, help = "the name (or id) of the release to update")]
    pub(crate) release: String,
    #[arg(long, short, help = projectHelp)]
    pub(crate) project: String,
}

#[derive(Parser, Clone)]
#[command()]
pub struct DeleteReleaseArgs {
    #[arg(long, short, help = releaseHelp)]
    pub(crate) release: String,
    #[arg(long, short, help = projectHelp)]
    pub(crate) project: String,
    #[arg(long, short, help = byIdHelp)]
    pub(crate) byId: Option<bool>,
    #[arg(long, help = "for tickets referencing this release, replace fix version with this other release")]
    pub(crate) replaceFixVersion: Option<String>,
    #[arg(long, help = "for tickets referencing this release, replace affected version with this other release")]
    pub(crate) replaceAffectedVersion: Option<String>,

}

impl DeleteReleaseArgs {
    fn set_release(&mut self, new: String) {
        self.release = new;
    }
    fn set_fixversion(&mut self, new: String) {
        self.replaceFixVersion = Option::from(new);
    }
    fn set_affectedversion(&mut self, new: String) {
        self.replaceAffectedVersion = Option::from(new);
    }
}

impl UpdateReleaseArgs {
    fn set_release(&mut self, new: String) {
        self.release = new;
    }
}

impl ListReleasesArgs {
    fn update_start_idx(&mut self, new: i64) {
        self.page_start_idx = new;
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Release {
    id: String,
    description: Option<String>,
    name: Option<String>,
    archived: bool,
    released: bool,
    releaseDate: Option<String>,
    overdue: Option<bool>,
    userReleaseDate: Option<String>,
    projectId: u64,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PaginatedStuff<T> {
    pub(crate) total: u64,
    pub(crate) startAt: u64,
    pub(crate) isLast: bool,
    pub(crate) values: Vec<T>,
}

pub fn execute_create_release(ctx: &Cli, args: &CreateReleaseArgs) -> Result<(), Box<dyn std::error::Error>> {
    let reqUrl = format!("{}/rest/api/3/version", ctx.baseJiraUrl, );
    let result = util::doPost::<Release, HashMap<&str, String>>(&reqUrl, ctx, &(assemble_create_args(args.clone())))?;
    Ok(())
}

pub fn execute_delete_release(ctx: &Cli, args: &DeleteReleaseArgs) -> Result<(), Box<dyn std::error::Error>> {
    let mut mutargs = args.clone();
    if args.byId.is_some() && args.byId.unwrap() == true {
        let reqUrl = format!("{}/rest/api/3/version/{}/removeAndSwap", ctx.baseJiraUrl, args.release);

        util::doPost::<(), HashMap<&str, String>>(&reqUrl, ctx, &(assemble_delete_args(args.clone())))?;
    } else {
        let id = get_id_from_name(ctx, args.project.clone(), args.release.clone())?;
        let reqUrl = format!("{}/rest/api/3/version/{}/removeAndSwap", ctx.baseJiraUrl, id);

        debug!("found release {} for name {}",id,args.release);
        mutargs.set_release(id);
        if args.replaceFixVersion.is_some() {
            let fixid = get_id_from_name(ctx, args.project.clone(), args.replaceFixVersion.unwrap().clone())?;
            mutargs.set_fixversion(fixid);
        }
        if args.replaceAffectedVersion.is_some() {
            let affectedid = get_id_from_name(ctx, args.project.clone(), args.replaceAffectedVersion.unwrap().clone())?;
            mutargs.set_fixversion(affectedid);
        }
        util::doPost::<(), HashMap<&str, String>>(&reqUrl, ctx, &(assemble_delete_args(mutargs.clone())))?;
    }

    Ok(())
}


pub fn execute_update_release(ctx: &Cli, args: &UpdateReleaseArgs) -> Result<(), Box<dyn std::error::Error>> {
    let mut mutargs = args.clone();
    if args.byId.is_some() && args.byId.unwrap() == true {
        let reqUrl = format!("{}/rest/api/3/version/{}", ctx.baseJiraUrl, args.release);

        util::doPut::<(), HashMap<&str, String>>(&reqUrl, ctx, &(assemble_update_args(args.clone())))?;
    } else {
        let id = get_id_from_name(ctx, args.project.clone(), args.release.clone())?;
        let reqUrl = format!("{}/rest/api/3/version/{}", ctx.baseJiraUrl, id);

        debug!("found release {} for name {}",id,args.release);
        mutargs.set_release(id);
        util::doPut::<(), HashMap<&str, String>>(&reqUrl, ctx, &(assemble_update_args(mutargs.clone())))?;
    }

    Ok(())
}

fn get_id_from_name(ctx: &Cli, project: String, name: String) -> Result<String, Box<dyn std::error::Error>> {
    let args = ListReleasesArgs {
        project,
        filter: Option::from(name),
        unpaginate: true,
        page_size: 100,
        page_start_idx: 0,
    };
    let mut result = do_list_releases(ctx, &args)?;
    if result.len() != 1 {
        Err(Box::try_from("release name matches multiple releases").unwrap())
    } else {
        let id = result.get(0).unwrap().id.clone();
        Ok(id)
    }
}

fn assemble_create_args(args: CreateReleaseArgs) -> HashMap<&'static str, String> {
    let mut params: HashMap<&str, String> = HashMap::new();
    params.insert("name", args.name);
    params.insert("project", args.project);
    if args.description.is_some() {
        params.insert("description", args.description.unwrap());
    }
    if args.startDate.is_some() {
        params.insert("startDate", args.startDate.unwrap());
    }
    if args.releaseDate.is_some() {
        params.insert("releaseDate", args.releaseDate.unwrap());
    }
    return params;
}

fn assemble_delete_args(args: DeleteReleaseArgs) -> HashMap<&'static str, String> {
    let mut params: HashMap<&str, String> = HashMap::new();
    if args.replaceAffectedVersion.is_some() {
        params.insert("moveAffectedIssuesTo", args.replaceAffectedVersion.unwrap());
    }
    if args.replaceFixVersion.is_some() {
        params.insert("moveFixIssuesTo", args.replaceFixVersion.unwrap());
    }

    return params;
}

fn assemble_update_args(args: UpdateReleaseArgs) -> HashMap<&'static str, String> {
    let mut params: HashMap<&str, String> = HashMap::new();
    if args.startDate.is_some() {
        params.insert("startDate", args.startDate.unwrap());
    }
    if args.releaseDate.is_some() {
        params.insert("releaseDate", args.releaseDate.unwrap());
    }
    if args.description.is_some() {
        params.insert("description", args.description.unwrap());
    }
    if args.name.is_some() {
        params.insert("name", args.name.unwrap());
    }
    if args.is_released.is_some() {
        params.insert("released", args.is_released.unwrap().to_string());
    }

    return params;
}

pub fn execute_list_releases(ctx: &Cli, args: &ListReleasesArgs) -> Result<(), Box<dyn std::error::Error>> {
    let values = do_list_releases(ctx, args)?;
    util::formatPrint::<Release>(values, ctx.output_format)?;


    Ok(())
}

fn do_list_releases(ctx: &Cli, args: &ListReleasesArgs) -> Result<Vec<Release>, Box<dyn std::error::Error>> {
    let (reqUrl, queryParams) = assemble_query(ctx, args);
    let url = reqUrl.clone();
    let mut values: Vec::<Release>;
    values = vec![];
    let mut res = util::doGet::<PaginatedStuff<Release>, HashMap<&str, String>>(&reqUrl, ctx, queryParams)?;

    if args.unpaginate {
        let arsgs = page_loop(res.total as i64, args);
        for arg in arsgs {
            let mut res = util::doGet::<PaginatedStuff<Release>, HashMap<&str, String>>(&url, ctx, arg)?;
            values.append(&mut res.values);
        }
    } else {
        values.append(&mut res.values);
    }
    Ok(values)
}

fn page_loop(totalResults: i64, args: &ListReleasesArgs) -> Vec<HashMap<&str, String>> {
    let remainder = totalResults % args.page_size;
    let mut pages = totalResults / args.page_size;
    let mut startidx = args.page_start_idx;
    if remainder > 0 {
        pages += 1;
    }
    let mut requests: Vec::<HashMap<&str, String>>;
    requests = vec![];
    for x in 0..pages {
        let mut newargs = args.clone();
        newargs.update_start_idx(startidx);
        requests.push(args_to_query_params(newargs));
        startidx = args.page_size + startidx;
    }
    return requests;
}

fn args_to_query_params(args: ListReleasesArgs) -> HashMap<&'static str, String> {
    let mut queryParams = HashMap::<&str, String>::new();
    queryParams.insert("startAt", args.page_start_idx.to_string());
    queryParams.insert("maxResults", args.page_size.to_string());
    if args.filter.is_some() {
        queryParams.insert("query", (*args.filter.as_ref().unwrap().clone().to_string()).parse().unwrap());
    }
    return queryParams.clone();
}

fn assemble_query<'a>(ctx: &Cli, args: &'a ListReleasesArgs) -> (String, HashMap<&'a str, String>) {
    let margs = args.clone();
    let reqUrl = format!("{}/rest/api/3/project/{}/version", ctx.baseJiraUrl, args.project);

    let queryParams = args_to_query_params(margs);
    return (reqUrl.clone(), queryParams.clone());
}


