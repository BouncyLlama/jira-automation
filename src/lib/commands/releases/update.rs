use std::collections::HashMap;
use crate::Cli;
use crate::lib::util;
use clap::Parser;
use super::*;


#[derive(Parser, Clone)]
#[command()]
pub struct UpdateReleaseArgs {
    #[arg(long, short, help = NAME_HELP)]
    pub(crate) name: Option<String>,
    #[arg(long, short, help = DESCRIPTION_HELP)]
    pub(crate) description: Option<String>,
    #[arg(long, help = START_DATE_HELP)]
    pub(crate) start_date: Option<String>,
    #[arg(long, help = RELEASE_DATE_HELP)]
    pub(crate) release_date: Option<String>,
    #[arg(long, short, help = "whether or not this release has been 'released'")]
    pub(crate) is_released: Option<bool>,
    #[arg(long, short, help = BY_ID_HELP)]
    pub(crate) by_id: Option<bool>,
    #[arg(long, short, help = "the name (or id) of the release to update")]
    pub(crate) release: String,
    #[arg(long, short, help = PROJECT_HELP)]
    pub(crate) project: String,
}


impl UpdateReleaseArgs {
    fn set_release(&mut self, new: String) {
        self.release = new;
    }
}

pub fn execute_update_release(ctx: &Cli, args: &UpdateReleaseArgs) -> Result<(), Box<dyn std::error::Error>> {
    let mut mutargs = args.clone();
    if args.by_id.is_some() && args.by_id.unwrap() == true {
        let req_url = format!("{}/rest/api/3/version/{}", ctx.base_jira_url, args.release);

        util::do_put::<(), HashMap<&str, String>>(&req_url, ctx, &(assemble_update_args(args.clone())))?;
    } else {
        let id = get_id_from_name(ctx, args.project.clone(), args.release.clone())?;
        let req_url = format!("{}/rest/api/3/version/{}", ctx.base_jira_url, id);

        debug!("found release {} for name {}",id,args.release);
        mutargs.set_release(id);
        util::do_put::<(), HashMap<&str, String>>(&req_url, ctx, &(assemble_update_args(mutargs.clone())))?;
    }

    Ok(())
}

fn assemble_update_args(args: UpdateReleaseArgs) -> HashMap<&'static str, String> {
    let mut params: HashMap<&str, String> = HashMap::new();
    if args.start_date.is_some() {
        params.insert("startDate", args.start_date.unwrap());
    }
    if args.release_date.is_some() {
        params.insert("releaseDate", args.release_date.unwrap());
    }
    if args.description.is_some() {
        params.insert("description", args.description.unwrap());
    }
    if args.name.is_some() {
        params.insert("name", args.name.unwrap());
    }
    if args.is_released.is_some() {
        params.insert("released", args.is_released.unwrap().to_string());
    }

    return params;
}