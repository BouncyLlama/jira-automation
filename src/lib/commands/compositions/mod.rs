use super::*;
use crate::lib::commands::issues::{
    do_search_issues, execute_update_issue, SearchIssuesArgs, UpdateIssueArgs,
};
use crate::lib::commands::releases::{
    do_update_release, CreateReleaseArgs, UpdateReleaseArgs,
};
use crate::lib::AppError;
use crate::lib::AppError::CouldNotCreateRelease;
use crate::Cli;
use chrono::{DateTime, Utc};
use clap::Parser;
use log::warn;
use releases::DESCRIPTION_HELP;
use releases::NAME_HELP;
use releases::PROJECT_HELP;

#[derive(Parser, Clone)]
#[command()]
pub struct ReleaseCompletedIssuesArgs {
    #[arg(long, short, help = NAME_HELP)]
    pub release_name: String,

    #[arg(long, short, help = PROJECT_HELP)]
    pub project: String,

    #[arg(long, short, help = "JQL query to determine which issues to release")]
    pub issue_jql: Option<String>,

    #[arg(long, short, help = DESCRIPTION_HELP)]
    pub description: Option<String>,
}

pub fn execute_do_release(ctx: &Cli, args: &ReleaseCompletedIssuesArgs) -> Result<(), AppError> {
    do_release(ctx, args)
}

pub fn do_release(ctx: &Cli, args: &ReleaseCompletedIssuesArgs) -> Result<(), AppError> {
    let now: DateTime<Utc> = Utc::now();
    let jql = match args.issue_jql.clone() {
        None => {
            format!(
                "(project = {0} AND status = Done) and (fixVersion is EMPTY)",
                args.project.clone()
            )
        }
        Some(s) => s,
    };

    let issuelist = do_search_issues(
        ctx,
        &SearchIssuesArgs {
            jql,
            unpaginate: true,
            page_size: 50,
            page_start_idx: 0,
        },
    )?;
    if issuelist.is_empty() {
        return Err(AppError::NoIssuesFound);
    }
    let releaseresult = releases::do_create_release(
        ctx,
        &CreateReleaseArgs {
            project: args.project.clone(),
            name: args.release_name.clone(),
            description: args.description.clone(),
            start_date: None,
            release_date: Some(now.format("%Y-%m-%d").to_string()),
        },
    )?;
    match releaseresult {
        Some(release) => {
            issuelist.iter().for_each(|issue| {
                match execute_update_issue(
                    ctx,
                    &UpdateIssueArgs {
                        name: issue.key.clone(),
                        fix_version: Some(release.id.clone()),
                        related_version: None,
                        use_version_id: true,
                    },
                ) {
                    Ok(_) => {}
                    Err(_) => {
                        warn!("could not add fix version to issue {0}", issue.key.clone())
                    }
                }
            });
            do_update_release(
                ctx,
                &UpdateReleaseArgs {
                    name: None,
                    description: None,
                    start_date: None,
                    release_date: None,
                    is_released: Some(true),
                    by_id: Some(true),
                    release: release.id.clone(),
                    project: args.project.clone(),
                },
            )?;
            Ok(())
        }
        _ => Err(CouldNotCreateRelease),
    }
}
