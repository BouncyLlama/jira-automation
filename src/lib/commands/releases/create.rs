use std::collections::HashMap;
use crate::Cli;
use crate::lib::commands::releases::{PaginatedReleases, Release};
use crate::lib::util;
use clap::Parser;
use super::*;

#[derive(Parser, Clone)]
#[command()]
pub struct CreateReleaseArgs {
    #[arg(long, short, help = projectHelp)]
    pub(crate) project: String,
    #[arg(long, short, help = nameHelp)]
    pub(crate) name: String,
    #[arg(long, short, help = descriptionHelp)]
    pub(crate) description: Option<String>,
    #[arg(long, short, help = startDateHelp)]
    pub(crate) startDate: Option<String>,
    #[arg(long, short, help = releaseDateHelp)]
    pub(crate) releaseDate: Option<String>,
}

fn assemble_create_args(args: CreateReleaseArgs) -> HashMap<&'static str, String> {
    let mut params: HashMap<&str, String> = HashMap::new();
    params.insert("name", args.name);
    params.insert("project", args.project);
    if args.description.is_some() {
        params.insert("description", args.description.unwrap());
    }
    if args.startDate.is_some() {
        params.insert("startDate", args.startDate.unwrap());
    }
    if args.releaseDate.is_some() {
        params.insert("releaseDate", args.releaseDate.unwrap());
    }
    return params;
}

pub fn execute_create_release(ctx: &Cli, args: &CreateReleaseArgs) -> Result<(), Box<dyn std::error::Error>> {
    let reqUrl = format!("{}/rest/api/3/version", ctx.baseJiraUrl, );
    let result = util::doPost::<Release, HashMap<&str, String>>(&reqUrl, ctx, &(assemble_create_args(args.clone())))?;
    Ok(())
}
