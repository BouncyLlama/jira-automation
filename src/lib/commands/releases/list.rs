use super::*;
use crate::lib::commands::releases::{PaginatedReleases, Release};
use crate::lib::util;
use crate::Cli;
use clap::Parser;
use std::collections::HashMap;

#[derive(Parser, Clone)]
#[command()]
pub struct ListReleasesArgs {
    #[arg(long, short, help = projectHelp)]
    pub(crate) project: String,
    #[arg(
    long,
    short,
    help = "optionally filter results; substring match in the name and description fields"
    )]
    pub(crate) filter: Option<String>,
    #[arg(
    long,
    short,
    default_value_t = false,
    help = "automatically query until all pages have been obtained"
    )]
    pub(crate) unpaginate: bool,
    #[arg(long, default_value_t = 50, help = "how many items to return")]
    pub(crate) page_size: i64,
    #[arg(long, default_value_t = 0, help = "item index to begin paging at")]
    pub(crate) page_start_idx: i64,
}

impl ListReleasesArgs {
    fn update_start_idx(&mut self, new: i64) {
        self.page_start_idx = new;
    }
}

pub fn execute_list_releases(
    ctx: &Cli,
    args: &ListReleasesArgs,
) -> Result<(), Box<dyn std::error::Error>> {
    let values = do_list_releases(ctx, args)?;
    util::formatPrint::<Release>(values, ctx.output_format)?;

    Ok(())
}

pub(crate) fn do_list_releases(
    ctx: &Cli,
    args: &ListReleasesArgs,
) -> Result<Vec<Release>, Box<dyn std::error::Error>> {
    let (reqUrl, queryParams) = assemble_query(ctx, args);
    let url = reqUrl.clone();
    let mut values: Vec<Release>;
    values = vec![];
    let mut res = util::doGet::<PaginatedReleases<Release>, HashMap<&str, String>>(&reqUrl, ctx, queryParams)?;

    if args.unpaginate {
        let arsgs = page_loop(res.total as i64, args);
        for arg in arsgs {
            let mut res = util::doGet::<PaginatedReleases<Release>, HashMap<&str, String>>(&url, ctx, arg)?;
            values.append(&mut res.values);
        }
    } else {
        values.append(&mut res.values);
    }
    Ok(values)
}

fn assemble_query<'a>(ctx: &Cli, args: &'a ListReleasesArgs) -> (String, HashMap<&'a str, String>) {
    let margs = args.clone();
    let reqUrl = format!("{}/rest/api/3/project/{}/version", ctx.baseJiraUrl, args.project);

    let queryParams = args_to_query_params(margs);
    return (reqUrl.clone(), queryParams.clone());
}

fn args_to_query_params(args: ListReleasesArgs) -> HashMap<&'static str, String> {
    let mut queryParams = HashMap::<&str, String>::new();
    queryParams.insert("startAt", args.page_start_idx.to_string());
    queryParams.insert("maxResults", args.page_size.to_string());
    if args.filter.is_some() {
        queryParams.insert("query", (*args.filter.as_ref().unwrap().clone().to_string()).parse().unwrap());
    }
    return queryParams.clone();
}


fn page_loop(totalResults: i64, args: &ListReleasesArgs) -> Vec<HashMap<&str, String>> {
    let remainder = totalResults % args.page_size;
    let mut pages = totalResults / args.page_size;
    let mut startidx = args.page_start_idx;
    if remainder > 0 {
        pages += 1;
    }
    let mut requests: Vec::<HashMap<&str, String>>;
    requests = vec![];
    for x in 0..pages {
        let mut newargs = args.clone();
        newargs.update_start_idx(startidx);
        requests.push(args_to_query_params(newargs));
        startidx = args.page_size + startidx;
    }
    return requests;
}

#[cfg(test)]
mod tests {
    use super::super::*;
    use super::*;

    use super::super::util::*;
    use crate::lib::util::Format;
    use crate::Cli;
    use httptest::{matchers::*, responders::*, Expectation, Server, ServerPool};
    use serde_json::json;
    use std::collections::HashMap;

    static SERVER_POOL: ServerPool = ServerPool::new(2);


    #[test]
    fn list_releases_paginated() -> Result<(), Box<dyn std::error::Error>> {
        let server = SERVER_POOL.get_server();

        let ctx = Cli {
            authToken: "".to_string(),

            output_format: Format::csv,
            userEmail: "".to_string(),
            baseJiraUrl: server.url("").to_string(),
            command: None,
        };
        let args = ListReleasesArgs {
            project: "foo".to_string(),
            filter: None,
            unpaginate: false,
            page_size: 1,
            page_start_idx: 0,
        };
        let resp = PaginatedReleases {
            total: 2,
            startAt: 0,
            isLast: true,
            values: vec![
                Release {
                    id: "1".to_string(),
                    description: None,
                    name: None,
                    archived: false,
                    released: false,
                    releaseDate: None,
                    overdue: None,
                    userReleaseDate: None,
                    projectId: 0,
                },
                Release {
                    id: "2".to_string(),
                    description: None,
                    name: None,
                    archived: false,
                    released: false,
                    releaseDate: None,
                    overdue: None,
                    userReleaseDate: None,
                    projectId: 0,
                },
            ],
        };
        // Start a server running on a local ephemeral port.
        // Configure the server to expect a single GET /foo request and respond
        // with a 200 status code.
        server.expect(
            Expectation::matching(any()).respond_with(json_encoded(serde_json::json!(resp))),
        );

        let res = do_list_releases(&ctx, &args).unwrap();
        insta::assert_debug_snapshot!(res);
        Ok(())
    }

    #[test]
    fn list_releases_unpaginated() -> Result<(), Box<dyn std::error::Error>> {
        let server = SERVER_POOL.get_server();

        let ctx = Cli {
            authToken: "".to_string(),

            output_format: Format::csv,
            userEmail: "".to_string(),
            baseJiraUrl: server.url("").to_string(),
            command: None,
        };
        let args = ListReleasesArgs {
            project: "foo".to_string(),
            filter: None,
            unpaginate: true,
            page_size: 1,
            page_start_idx: 0,
        };
        let resp1 = PaginatedReleases {
            total: 2,
            startAt: 0,
            isLast: false,
            values: vec![
                Release {
                    id: "1".to_string(),
                    description: None,
                    name: None,
                    archived: false,
                    released: false,
                    releaseDate: None,
                    overdue: None,
                    userReleaseDate: None,
                    projectId: 0,
                },
            ],
        };
        let resp2 = PaginatedReleases {
            total: 2,
            startAt: 1,
            isLast: true,
            values: vec![
                Release {
                    id: "2".to_string(),
                    description: None,
                    name: None,
                    archived: false,
                    released: false,
                    releaseDate: None,
                    overdue: None,
                    userReleaseDate: None,
                    projectId: 0,
                },
            ],
        };
        // Start a server running on a local ephemeral port.
        // Configure the server to expect a single GET /foo request and respond
        // with a 200 status code.
        server.expect(
            Expectation::matching(request::query(url_decoded(contains(("startAt", "0"))))).times(2)
                .respond_with(json_encoded(serde_json::json!(resp2)))
        );
        server.expect(Expectation::matching(request::query(url_decoded(contains(("startAt", "1"))))).respond_with(json_encoded(serde_json::json!(resp1))));

        // server.expect(
        //     Expectation::matching(request::query(url_decoded(contains(("startAt", any()))))).respond_with(json_encoded(serde_json::json!(resp2))), );

        let res = do_list_releases(&ctx, &args)?;

        insta::assert_debug_snapshot!(res);
        Ok(())
    }
}
