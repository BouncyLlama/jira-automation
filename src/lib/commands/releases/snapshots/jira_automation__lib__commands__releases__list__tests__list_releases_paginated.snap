---
source: src/lib/commands/releases/list.rs
expression: res
---
[
    Release {
        id: "1",
        description: None,
        name: None,
        archived: false,
        released: false,
        releaseDate: None,
        overdue: None,
        userReleaseDate: None,
        projectId: 0,
    },
    Release {
        id: "2",
        description: None,
        name: None,
        archived: false,
        released: false,
        releaseDate: None,
        overdue: None,
        userReleaseDate: None,
        projectId: 0,
    },
]
